### PR TITLE
Repairing non-2-manifold mesh

### PR DESCRIPTION
**Pull request by katarinagresova**
Adding `removeNon2ManifoldParts()` functionality to `repairRoofMesh()`.

### CODE REVIEW

File: semantic/mesh/roof/mesh-repair.cpp [hunk 1]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+    std::set<Face::index_type> nonManifoldfaces;
+    for (auto& edge : edgeMap) {
+        if (edge.second.size() > 2) {
+            for (auto& fi : edge.second) {
+                nonManifoldfaces.insert(fi);
+            }
+        }
+    }
+
+    if ((int)nonManifoldfaces.size() == 0) { return mesh; }
```
Comment (vaclavblazek) [inline semantic/mesh/roof/mesh-repair.cpp]: Pls, no C-cast. Either use _ctor_ cast like `int(...)` or full blown `static_cast` like `static_cast<int>(...)`. Not ti mention that there's no need to cast when comparing with zero.

However, something completely different belongs here:
```
if (nonManifoldfaces.empty()) { return mesh; }
```

File: semantic/mesh/roof/mesh-repair.cpp [hunk 2]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+}
+
+
+geometry::Mesh removeNon2ManifoldParts(const geometry::Mesh& mesh)
+{
+    geometry::EdgeMap edgeMap = getNonManifoldEdgeMap(mesh);
+
+    // collect faces incident with non-manifold edge
+    std::set<Face::index_type> nonManifoldfaces;
+    for (auto& edge : edgeMap) {
```
Comment (vaclavblazek) [inline semantic/mesh/roof/mesh-repair.cpp]: edge is not modified -> `const`

File: semantic/mesh/roof/mesh-repair.cpp [hunk 3]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+        if (edge.second.size() > 2) {
+            for (auto& fi : edge.second) {
```
Comment (vaclavblazek) [inline semantic/mesh/roof/mesh-repair.cpp]: `const` as well

File: semantic/mesh/roof/mesh-repair.cpp [hunk 4]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
 }
 
+
+void connectedFaces(const geometry::Mesh& mesh,
+                    const geometry::EdgeMap& edgeMap,
+                    const std::set<Face::index_type>& nonManifoldfaces,
+                    std::vector<int>& regions,
+                    std::vector<int>& validRegions)
+{
+
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp]: Just a cosmetic detail
```suggestion
```

File: semantic/mesh/roof/mesh-repair.cpp [hunk 5]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+        ++comp;
+    }
+}
+
+
+geometry::Mesh constructMeshPart(const geometry::Mesh& mesh, 
+                                 const std::vector<int>& regions, 
+                                 const int regionId)
+{
+
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp:304]:
```suggestion
```

File: semantic/mesh/roof/mesh-repair.cpp [hunk 6]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+            for (const auto& e : triEdges) {
+                // skip non-manifold edges
+                if (edgeMap.at(e).size() > 2) { continue; }
+
+                if (edgeMap.at(e).size() < 2) {
+                    LOG(info1) << "Edge has only one face."
+                    << " Marking component " << comp << " as non-manifold.";
+                    isValid = false;
+                }
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp]: I'd note here that this is a watertightness condition for the new component. A manifold can have a boundary and still be a 2-manifold. Maybe even issue a warning that we got non-watertight mesh on the input.

File: semantic/mesh/roof/mesh-repair.cpp [hunk 7]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+            vertexIndexMap[b] = vertexId;
+            vertexId += 1;
+        }
+
+        math::Point3 c = mesh.c(face);
+        if (!vertexIndexMap.count(c)) {
+            meshPart.vertices.push_back(c);
+            vertexIndexMap[c] = vertexId;
+            vertexId += 1;
+        }
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp]: Same code repeated three times. Maybe better to create a separate function or lambda - sth like:
```
auto getPtIdx = [&](const math::Point3& a){
   if (!vertexIndexMap.count(a)) {
      meshPart.vertices.push_back(a);
      vertexIndexMap[a] = vertexId;
      vertexId += 1;
   }
   return vertexIndexMap[a];
};

meshPart.addFace(getPtIdx(a), getPtIdx(b), getPtIdx(c), face.imageId);
```

File: semantic/mesh/roof/mesh-repair.cpp [hunk 8]
```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
...
+
+
+geometry::Mesh getValidComponents(const geometry::Mesh& mesh,
+                                  const std::vector<int>& regions,
+                                  const std::vector<int>& validRegions)
+{
+    geometry::Mesh validMesh;
+    for (int regionId : validRegions) {
+        geometry::Mesh meshPart(constructMeshPart(mesh, regions, regionId));
+        if (meshPart.volume() < 0.0001) {
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp]: Introduce as named constant pls

File: semantic/mesh/roof/mesh-repair.cpp [hunk 9]
```diff
@@ -242,6 +394,7 @@ Mesh removeUnusedVertices(Mesh& mesh)
  * - removes faces that are adjacent to each other from all three sides
  * - removes null faces (collinear vertices) by flipping their longest edge
  * - removes unused vertices
+ * - removes non 2-manifold parts
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp:389]: I'd rather add a note that this function requires watertight mesh

**Review by tomas2211 — COMMENTED** [review]
Sorry I got to the review a bit later. Good job overall.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 1 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: You are right,  changed.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 2 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: Added.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 3 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: Added.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 5 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp:304]: Removed.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 7 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: Added.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 8 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: Done.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 6 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp]: Changed to warning and mentioned non-watertight mesh.

File: semantic/mesh/roof/mesh-repair.cpp (hunk 9 above)
Comment (katarinagresova) [reply semantic/mesh/roof/mesh-repair.cpp:389]: Added

**Review by tomas2211 — APPROVED** [review]

**Review by vaclavblazek — APPROVED** [review]
seems ok

File: semantic/mesh/roof/mesh-repair.cpp [hunk 10]
```diff
@@ -25,12 +25,17 @@
...
 
 #include "dbglog/dbglog.hpp"
 #include "geometry/mesh.hpp"
 #include "geometry/meshop.hpp"
 
+#include "../../mesh.hpp"
+#include "../detail.hpp"
+
 #include "math/geometry.hpp"
+#include "vadstena-libs/meshio.hpp"
```
Comment (tomas2211) [inline semantic/mesh/roof/mesh-repair.cpp:38]: I guess this obsolete include of an internal library slipped through. I hope it's not used anywhere - please check and fix @katarinagresova

**Pull request merged**